# seabreeze
_seabreeze_ is a tool for the comprehensive analysis of structural variant (SV) mutations in bacterial resequencing experiments; it automates and unites several analyses into a single, easy-to-use pipeline. _seabreeze_ uses a combination of existing packages and custom scripts, with Snakemake to manage dependencies and the workflow.

- To get started, please refer to the [installation](installation.md).
- For an overview of the commands, please see [usage](usage.md). Each command can generate several output files, however most users will usually only find a subset of them relevant. These have been highlighted in the [usage](usage.md) itself. A more extensive documentation of the other output files is on the [output](output.md) page. Wherever the output corresponds to a file type/format from an existing package, the user is redirected to the respective package's documentation. Wherever the output corresponds a new file format specific to _seabreeze_, it is described on the [output](output.md) page.
- If you would like to learn how to use _seabreeze_, give the [tutorial](tutorial.md) a try.
- _seabreeze_ is available under the MIT License, and source code is on [GitHub](https://github.com/barricklab/seabreeze).
- If you have any questions, want to report bugs/error or submit a request for a new feature, please start a new [GitHub issue](https://github.com/barricklab/seabreeze/issues). 
- See other software to analyse bacterial genomes from the [Barrick Lab at the University of Texas at Austin](https://barricklab.org/twiki/bin/view/Lab/SoftwareList. )
- Contact the creators at irazibbu@utexas.edu or jbarrick@cm.utexas.edu.

