# Output

## Analyse genome sizes

This command generates `data/04_rename_genome/genome_size_stats.csv` as output. This csv file has the following fields:

- assembly: name of the assembly, as specified in `data/data.csv`
- ancestor: name of the ancestor, as specified in `data/data.csv`
- size_assembly: size of the assembly, in bp
- size_ancestor: size of the ancestor, in bp
- difference: size_assembly - size_ancestor. This value is negative
- percent_change: the difference in size, as a percent of the size of the assembly 

Here is a sample table generate from the [tutorial](tutorial.md)

|assembly|ancestor|size_assembly|size_ancestor|difference|percent_change|
|-----|-------|------|-------|----|----|
|REL606|REL606|4629812|4629812|0|0.0|
|REL606_evolved_1|REL606|4617111|4629812|-12701|-0.2743|
|REL606_evolved_2|REL606|4549910|4629812|-79902|-1.7258|
|REL606_evolved_3|REL606|4673209|4629812|43397|0.9373|

## Predict SVs

[mummer4](https://github.com/mummer4/mummer) is used to perform alignments between assembly-ancestor pairs. The alignment files are stored in `data/06_nucmer_alignment/<assembly>/` directories. Each of these directories contains the following files:

- `<assembly>.delta`: The all-vs-all alignment generated by the `nucmer` command of _mummer4_ 
- `<assembly>.delta.filter`: The above file is filtered by the `delta-filter` command to only retain alignments > 95% sequence identity and > 100bp long
- `<assembly>.filtered.coords`: A file summarising information about each delta alignment in the filtered file.

For specific information about these file formats, please refer to the [mummer manual](https://mummer.sourceforge.net/manual/#coords). 

Next [SyRI](https://github.com/schneebergerlab/syri) is used to call SVs using those alignments. The output files for each assembly is in the respective `data/07_syri_output/<assembly>/` directories. Each of these directories contains the following files:

- `<assembly>syri.out`: a tsv file that describes the location and type of SVs in the assembly, relative to its ancestor. We refer the user to the official [SyRI documentation](https://schneebergerlab.github.io/syri/fileformat.html) for information on interpreting this table.
- `<assembly>syri.vcf`: a variant calling format (vcf) file describing the same SVs

Raw output generated by [SyRI](https://github.com/schneebergerlab/syri) was refined with custom python scripts to predict a minimal set of structural variants. Briefly, new insertions of individual IS elements due to transposition events were filtered out, ambiguous nomenclature for deletions was replaced with consistent terminology, falsely classified duplications were corrected, adjacent deletions were merged, and adjacent syntenic regions were merged. This filtered set of SVs is a tsv file `data/07_syri_output/<assembly>/<assembly>_clean.syri.out` and follows the same file format as `<assembly>syri.out`. 

Finally, a modified version of [plotsr](https://github.com/ira-zibbu/plotsr) is used to generate synteny plots. Two synteny plots are generated in `data/07_syri_output/<assembly>/`:

- `<assembly>.plot.pdf`: synteny plot corresponding to `<assembly>syri.out`
- `<assembly>.plot.2.pdf`: synteny plot corresponding to `<assembly>_clean.syri.out`
